### ISSUE TITLE
torch::max: CPU scalar_tensor and CUDA tensor causes ...is_cuda() INTERNAL ASSERT FAILED at ".../Loops.cuh":94, please report a bug to PyTorch.

### ISSUE THREAD

**Issue by mreppen** [pytorch/pytorch#51013]
## 🐛 Bug

Using a CPU `scalar_tensor` with `torch::max` and a CUDA tensor produces
`iter.device(arg).is_cuda() INTERNAL ASSERT FAILED at "/build/python-pytorch/src/pytorch-1.7.1-opt-cuda/aten/src/ATen/native/cuda/Loops.cuh":94, please report a bug to PyTorch.`

Other functions, like `torch::add` do not suffer from the same problem.

## To Reproduce

1. Create CUDA tensor
2. Create scalar_tensor
3. Compute `torch::max`

(of course, if explicitly specifying `kCUDA` for the `scalar_tensor` below, it works again)

```cpp
#include <torch/torch.h>
#include <iostream>

int main() {
  torch::Tensor x  = torch::scalar_tensor({2.0});
  torch::Tensor y = torch::ones({3,1}, torch::kCUDA);
  std::cout << torch::add(x, y) << std::endl;
  std::cout << torch::max(x, y) << std::endl;
}
```
```
 3
 3
 3
[ CUDAFloatType{3,1} ]
terminate called after throwing an instance of 'c10::Error'
  what():  iter.device(arg).is_cuda() INTERNAL ASSERT FAILED at "/build/python-pytorch/src/pytorch-1.7.1-opt-cuda/aten/src/ATen/native/cuda/Loops.cuh":94, please report a bug to PyTorch.
Exception raised from gpu_kernel at /build/python-pytorch/src/pytorch-1.7.1-opt-cuda/aten/src/ATen/native/cuda/Loops.cuh:94 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x72 (0x7f78838a87c2 in /usr/lib/libc10.so)
frame #1: <unknown function> + 0x97fc38 (0x7f78233c3c38 in /usr/lib/libtorch_cuda.so)
frame #2: <unknown function> + 0x118d448 (0x7f787929a448 in /usr/lib/libtorch_cpu.so)
frame #3: at::native::maximum(at::Tensor const&, at::Tensor const&) + 0xa1 (0x7f787928ee71 in /usr/lib/libtorch_cpu.so)
frame #4: <unknown function> + 0x2f32ba2 (0x7f7825976ba2 in /usr/lib/libtorch_cuda.so)
frame #5: <unknown function> + 0x2f63986 (0x7f78259a7986 in /usr/lib/libtorch_cuda.so)
frame #6: <unknown function> + 0x17165b0 (0x7f78798235b0 in /usr/lib/libtorch_cpu.so)
frame #7: at::maximum(at::Tensor const&, at::Tensor const&) + 0xb6 (0x7f7879824226 in /usr/lib/libtorch_cpu.so)
frame #8: <unknown function> + 0x2a85b48 (0x7f787ab92b48 in /usr/lib/libtorch_cpu.so)
frame #9: <unknown function> + 0x102c7b6 (0x7f78791397b6 in /usr/lib/libtorch_cpu.so)
frame #10: <unknown function> + 0x17165b0 (0x7f78798235b0 in /usr/lib/libtorch_cpu.so)
frame #11: at::maximum(at::Tensor const&, at::Tensor const&) + 0xb6 (0x7f7879824226 in /usr/lib/libtorch_cpu.so)
frame #12: at::native::max(at::Tensor const&, at::Tensor const&) + 0x1f (0x7f787928840f in /usr/lib/libtorch_cpu.so)
frame #13: <unknown function> + 0x17d556c (0x7f78798e256c in /usr/lib/libtorch_cpu.so)
frame #14: <unknown function> + 0x102c7b6 (0x7f78791397b6 in /usr/lib/libtorch_cpu.so)
frame #15: <unknown function> + 0x17165b0 (0x7f78798235b0 in /usr/lib/libtorch_cpu.so)
frame #16: at::max(at::Tensor const&, at::Tensor const&) + 0xb6 (0x7f78798279f6 in /usr/lib/libtorch_cpu.so)
frame #17: main + 0x140 (0x55bedd9d56d2 in ./test)
frame #18: __libc_start_main + 0xf2 (0x7f78226ac152 in /usr/lib/libc.so.6)
frame #19: _start + 0x2e (0x55bedd9d547e in ./test)
```

## Expected behavior

`torch::max` should work if `torch::add` does

## Environment

PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 11.2
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 10.2.0
Clang version: Could not collect
CMake version: version 3.19.3

Python version: 3.9 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1650 Ti
Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.0.5
/usr/lib/libcudnn_adv_infer.so.8.0.5
/usr/lib/libcudnn_adv_train.so.8.0.5
/usr/lib/libcudnn_cnn_infer.so.8.0.5
/usr/lib/libcudnn_cnn_train.so.8.0.5
/usr/lib/libcudnn_ops_infer.so.8.0.5
/usr/lib/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.7.1
[conda] Could not collect

## Additional context

For reference: https://github.com/LaurentMazare/ocaml-torch/issues/53

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @ngimel

**Comment by mruberry** [comment]
We would accept a PR fixing this.

**Comment by ngimel** [comment]
Can't repro on 1.7 and on master. Here's python repro script (passes, prints things as expected, ` self` argument is unused, as in original script). cpp extension passes too. 
```

cpp_source = """
    at::Tensor max_helper(at::Tensor self) {
    torch::Tensor y = torch::ones({3,1}, torch::kCUDA);
    torch::Tensor x  = torch::scalar_tensor({2.0});
    std::cout << torch::add(x, y) << std::endl;
    auto out = torch::max(x,y);
    std::cout <<out << std::endl;
    return out;

  }
"""
import torch
from torch.utils.cpp_extension import load_inline
module = load_inline(name='ext', cpp_sources=[cpp_source], functions=['max_helper'])
inp1=torch.randn(2,2, device="cuda")
module.max_helper(inp1)
```

**Comment by ngimel** [comment]
It was fixed in #48446, so should be fixed on nightlies/master, 1.7.1 has this bug.

**Comment by mreppen** [comment]
After working against cmake for way too long I confirm this on master. Sorry about the noise.
